### PR TITLE
formula: include versioned `*.so.*` libs for `shared_library("*")`

### PR DESCRIPTION
**Pull request by SeekingMeaning**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?
- [ ] Have you successfully run `brew man` locally and committed any changes?

-----

`ipopt` and `libgsm` use `Formula#shared_library` to locate shared libraries:

```rb
lib.install_symlink Dir["#{libexec}/lib/#{shared_library("*")}"]
```

However, this only matches unversioned `*.so` files such as `libpord.so` but ignores versioned `*.so.*` files like `libpord.so.0`. This PR modifies `shared_library` so that it matches both.

Related:
- https://github.com/Homebrew/homebrew-core/pull/67553
- https://github.com/Homebrew/homebrew-core/search?q=shared_library
- https://github.com/Homebrew/homebrew-core/blob/master/Formula/ipopt.rb#L54-L58
- https://github.com/Homebrew/homebrew-core/blob/master/Formula/libgsm.rb#L48

### CODE REVIEW

**Review by iMichka — APPROVED** [review]
Nice !

File: Library/Homebrew/extend/os/linux/formula.rb [hunk 1]
```diff
@@ -5,6 +5,8 @@ class Formula
   undef shared_library
 
   def shared_library(name, version = nil)
+    return "*.so*" if name == "*" && (version.nil? || version == "*")
```
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/os/linux/formula.rb:8]:
```suggestion
    return "*.so*" if name == "*" && (version.blank? || version == "*")
```

File: Library/Homebrew/extend/os/linux/formula.rb [hunk 2]
```diff
@@ -5,6 +5,8 @@ class Formula
...
+
     "#{name}.so#{"." unless version.nil?}#{version}"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/os/linux/formula.rb:10]:
```suggestion
    "#{name}.so#{"." if version.present?}#{version}"
```

**Review by MikeMcQuaid — APPROVED** [review]
Nice catch! Some style suggestions, not blocking.

**Comment by iMichka** [comment]
If someone has technical information about why .so files can have version numbers after the extension that I can read up, I would be interested. Just for personal knowledge :)

**Comment by gromgit** [comment]
> If someone has technical information about why .so files can have version numbers after the extension that I can read up, I would be interested. Just for personal knowledge :)

A little light reading then. :laughing: 
* The LDP's [Program Library HOWTO, Chap. 3](https://tldp.org/HOWTO/Program-Library-HOWTO/shared-libraries.html) seems to be the most-cited reference, and is a readable overview
* Douglas Creager wrote [an ELI12 version](https://dcreager.net/shared-library-versions/), with brief cameos by libtool and CMake
* Ulrich Drepper's [How to Write Shared Librares [PDF]](https://akkadia.org/drepper/dsohowto.pdf), for when you have time for a deep dive without oxygen

**Comment by SeekingMeaning** [comment]
Updated to be more consistent, please review the tests to make sure I got the edge cases right

**Review by MikeMcQuaid — APPROVED** [review]

**Review by iMichka — APPROVED** [review]

**Pull request merged**